crates/oxide/src/lib.rs

```diff
@@ -427,35 +451,59 @@ impl Scanner {
 }
 
 fn read_changed_content(c: ChangedContent) -> Option<Vec<u8>> {
-    if let Some(content) = c.content {
-        return Some(content.into_bytes());
-    }
-
-    let Some(file) = c.file else {
-        return Default::default();
-    };
+    let (content, extension) = match c {
+        ChangedContent::File(file, extension) => match std::fs::read(&file) {
+            Ok(content) => (content, extension),
+            Err(e) => {
+                event!(tracing::Level::ERROR, "Failed to read file: {:?}", e);
+                return None;
+            }
+        },
 
-    let Ok(content) = std::fs::read(&file).map_err(|e| {
-        event!(tracing::Level::ERROR, "Failed to read file: {:?}", e);
-        e
-    }) else {
-        return Default::default();
+        ChangedContent::Content(contents, extension) => (contents.into_bytes(), extension),
     };
 
-    let Some(extension) = file.extension().map(|x| x.to_str()) else {
-        return Some(content);
-    };
+    Some(pre_process_input(&content, &extension))
+}
 
+pub fn pre_process_input(content: &[u8], extension: &str) -> Vec<u8> {
+    // NOTE: Replace strings with same-length replacements to avoid reallocations
     match extension {
-        // Angular class shorthand
-        Some("html") => Some(content.replace("[class.", "[")),
-        Some("svelte") => Some(
-            content
-                .replace(" class:", " ")
-                .replace("\tclass:", " ")
-                .replace("\nclass:", " "),
-        ),
-        _ => Some(content),
+        // Handle dot notation, e.g.: `div.flex.bg-red-500` -> `div flex bg-red-500`
+        "slim" | "pug" => {
+            let len = content.len();
+            let mut cursor = cursor::Cursor::new(content);
+            let mut new_content = String::from_utf8(content.to_vec()).unwrap();
+            let mut string_machine = StringMachine;
+
+            while cursor.pos < len {
+                match cursor.curr {
+                    // Consume strings as-is
+                    b'\'' | b'"' => {
+                        string_machine.next(&mut cursor);
+                    }
+
+                    // Replace dots with spaces
+                    b'.' => {
+                        new_content.replace_range(cursor.pos..cursor.pos + 1, " ");
+                    }
+
+                    // Consume everything else
+                    _ => {}
+                };
+
+                cursor.advance();
+            }
+
+            new_content.into()
+        }
```

Comment:
Didn't want to encode this logic in the state machines themselves. Because state machines run to completion which might include `.` characters already.

Preprocessing this also means that we don't pay an additional cost _if_ you are dealing with other types of extensions.

Reply:
Yeah, I like this implementation. I'm cool with having some languages pay a small up-front cost to keep everything else fast and maintainable.

Reply:
Yep that's definitely the way to go for these